### ISSUE TITLE
Hi, Thanks for you reply in stackoverflow

### ISSUE THREAD

**Issue by johnnykthink** [pltrdy/tf_rnnlm#1]
Hi pltrdy,

Thanks for you reply my question at: http://stackoverflow.com/questions/41096896/how-to-speedup-rnn-training-speed-of-tensorflow

I search the github the same id and found here, I think it's you ;-)

**Comment by pltrdy** [comment]
Indeed.
I'm replying on stack

**Comment by pltrdy** [comment]
I just contact you here because I've no other way to do, note that the dynamic_rnn is broken and actually cheats about its perplexity as a side effect of doing the batching wrong. I've inspected batched data and it turns out that we keep <eos> as the last word of `x`, which should not be the case.

I won't go into details here about how it make a better PPL, the fact is that once it is fixed the PPL gets bad. Around 130 for small PTB case, which is worst than initial script. Disappointing.

**Comment by johnnykthink** [comment]
@pltrdy Thanks for reply. My email is johnnykthink@gmail.com.
I had read your code and have a question:
`    elif fct == "sampledsoftmax":
      def _loss_fct(inputs_, labels_):
        labels_ = tf.reshape(labels_, [-1, 1])
        return tf.nn.sampled_softmax_loss(
            self.w, self.b, inputs_, labels_, self.num_samples, vocab_size)
    
      loss = tf.nn.seq2seq.sequence_loss_by_example(
            [self.output],
            [tf.reshape(self.targets, [-1,1])],
            [self.mask],
            softmax_loss_function=_loss_fct)`

And my code is: 
`loss = tf.nn.sampled_softmax_loss(softmax_w, softmax_b, output, self.targets, config.num_samples, config.vocab_size)`

What the difference between these?

And I found a strange things that the tf.nn.dynamic_rnn's predict accuray is much bad than seq2seq.rnn_decoder, the prediction of tf.nn.dynamic_rnn is not make sense, but seq2seq.rnn_decoder's prediction is impressive.

Do you have any ideas about these?

Thanks.

**Comment by pltrdy** [comment]
1) Using `tf.nn.seq2seq.sequence_loss_by_example` or directly `tf.nn.sampled_softmax_loss` may give the same result. It may be quite optimized using seq_loss_by_example but I'm not even sure. 
Just notice that I'm passing a `mask` vector (a with 0/1) to "ignore" some outputs in loss calculation (i'm ignoring padded elements)

2) I didn't used `rnn_decoder` and I'm not sure how I would do it if I wanted to. Tell me if you have any implementation that would give better results than mine. Try to run it using default setup (PTB Dataset, `small, medium, large` configurations as described in `config.py` to have a common baseline. (you can obviously share your own result and describe and/or link to the dataset involved.

**Comment by pltrdy** [comment]
In this case I would suggest you to start from wherever you prefer and go into a new branch e.g. `decode` that explore the use of decoder rnn for word lm. This would be a great contribution :)

**Comment by johnnykthink** [comment]
Hi, I had test the sampled softmax with sequence loss by example, but the finally loss is 10% bigger than the use sampled softmax only.

For rnn_decoder vs dynamic_rnn, I had increase the train epochs to 150, the prediction result is similar with rnn_decoder(50 epochs), and the result is make sense, but I don"t know why. (>_<)

For the padding of variable sequence, I found a strange thing of it, the question is here: [http://stackoverflow.com/questions/41235416/how-to-improve-the-word-rnn-accuracy-in-tensorflow](http://stackoverflow.com/questions/41235416/how-to-improve-the-word-rnn-accuracy-in-tensorflow), maybe you can give me some insight.

Thanks.

**Comment by johnnykthink** [comment]
Another question, why need tf.to_float here? 
I know the input data type is float, but the input data is all integer, why need convert to float here?

`_mask = tf.sign(tf.to_float(_inputs))`

**Comment by pltrdy** [comment]
1) `sequence loss by example` should not give different result. In the `dynamic` case you must pass a `mask` vector. You may expect incorrect perplexity if you dont

2) About `rnn_decoder`, I don't know as well, I'll keep this in mind. Interesting.

3) Lukasz Kaiser's answer about padding on stackoverflow is interesting. It contains key element to understand. 

4) It looks like `tf.sign` accepts `int32` so `to_float` is useless. Maybe did it before required a cast. Don't remember. Just ignore it :)

**Comment by johnnykthink** [comment]
Hi,  
I did some additional processing of the training set,:
> Original Title: This is a iPhone case
> I convert it to multi-lines, like:
> This is a iPhone case
> is a iPhone case
> a iPhone case
> iPhone case

But I think it's not a very good solution, because it's make the training time 5x.
It's has a perfect accuracy in my 10k lines test data that make me very impressed . But in 80m train data it's prediction is bad, I was debugging, maybe the problem is in hyperparams.
If I have any progress, I will tell you.

I have a question about num_samples of sampled_softmax_loss: 
My vocab_size is 450k and num_samples is 512. 

These two values differ too much, will it bring a big affect to the loss?


Thanks, Happy new year ;-)

**Comment by pltrdy** [comment]
You should try higher values for num_samples. It gives better results while not much slowing the process. I usually use 1024 as num_samples, as in our case we empirically found that it was good.

Keep an eye on updates on dynamic branch. Some fixes may be quite important as it is still experimental.

**Comment by johnnykthink** [comment]
Hi,
I had use num_samples 2880 in my training, it's give me a better accuracy.

And I found a strange things that the accuracy is begin to decrease after epoch 1, but the loss is still decreasing. The screenshot is here.
<img width="992" alt="screen shot 2017-02-09 at 10 18 55" src="https://cloud.githubusercontent.com/assets/2732807/22767046/ee792312-eeb3-11e6-9f1d-b35de4548cb3.png">

<img width="995" alt="screen shot 2017-02-09 at 10 19 14" src="https://cloud.githubusercontent.com/assets/2732807/22766743/33619eac-eeb2-11e6-95a8-f2df797f5d90.png">



So I doubt that my loss defintion is not correct. And I compare my code with yours, 
My code:
`
        def _loss_fct(inputs_, labels_):
            labels_ = tf.reshape(labels_, [-1, 1])
            return tf.nn.sampled_softmax_loss(
                softmax_w, softmax_b, inputs_, labels_, config.num_samples, config.vocab_size)

        loss = tf.nn.seq2seq.sequence_loss_by_example(
            [output],
            [tf.reshape(self.targets, [-1, 1])],
            [tf.reshape(mask, [-1])],
            softmax_loss_function=_loss_fct)

self.cost = tf.reduce_sum(loss) / batch_size / tf.reduce_sum(seq_len)`

Your code:
`_cost = tf.reduce_sum(loss) / batch_size`

In my code I divided the sum(seq_len), but I don't sure it's right.
So what do you think of it?
